control-plane/msp-operator/src/main.rs

```diff
@@ -0,0 +1,910 @@
+                }
+            },
+        )
+        .await
+        .map_err(|e| error!(?e));
+
+        Ok(ReconcilerAction {
+            requeue_after: None,
+        })
+    }
+}
+
+/// Determine what we want to do when dealing with errors from the
+/// reconciliation loop
+fn error_policy(error: &Error, _ctx: Context<OperatorContext>) -> ReconcilerAction {
+    let duration = Duration::from_secs(match error {
+        Error::Duplicate { timeout } | Error::SpecError { timeout, .. } => (*timeout).into(),
+
+        Error::ReconcileError { .. } => {
+            return ReconcilerAction {
+                requeue_after: None,
+            };
+        }
+        _ => 5,
+    });
+
+    let when = Utc::now()
+        .checked_add_signed(chrono::Duration::from_std(duration).unwrap())
+        .unwrap();
+    warn!(
+        "{}, retry scheduled @{} ({} seconds from now)",
+        error,
+        when.to_rfc2822(),
+        duration.as_secs()
+    );
+    ReconcilerAction {
+        requeue_after: Some(duration),
+    }
+}
+
+/// The main work horse
+#[tracing::instrument(fields(name = %msp.spec.node, status = ?msp.status) skip(msp, ctx))]
+async fn reconcile(
+    msp: MayastorPool,
+    ctx: Context<OperatorContext>,
+) -> Result<ReconcilerAction, Error> {
+    let ctx = ctx.into_inner();
+    let msp = ctx.upsert(ctx.clone(), msp).await;
+
+    let _ = msp.finalizer().await;
+
+    match msp.status {
+        Some(MayastorPoolStatus {
+            state: PoolState::Creating,
+            ..
+        }) => {
+            return msp.create_or_import().await;
+        }
+
+        Some(MayastorPoolStatus {
+            state: PoolState::Created,
+            ..
+        }) => {
+            return msp.online_pool().await;
+        }
+
+        Some(MayastorPoolStatus {
+            state: PoolState::Online,
+            ..
+        }) => {
+            return msp.pool_check().await;
+        }
+
+        Some(MayastorPoolStatus {
+            state: PoolState::Error,
+            ..
+        }) => {
+            error!(pool = ?msp.name(), "entered error as final state");
+            Err(Error::ReconcileError { name: msp.name() })
+        }
+
+        // We use this state to indicate its a new CRD howoever, we could (and
+        // perhaps should) use the finalizer callback.
+        None => return msp.start().await,
+    }
+}
+
+async fn pool_controller(args: ArgMatches<'_>) -> anyhow::Result<()> {
+    let k8s = Client::try_default().await?;
+    let namespace = args.value_of("namespace").unwrap();
+    let msp: Api<MayastorPool> = Api::namespaced(k8s.clone(), namespace);
+    let lp = ListParams::default();
+
+    let context = Context::new(OperatorContext {
+        k8s,
+        inventory: tokio::sync::RwLock::new(HashMap::new()),
+        url: Url::parse(args.value_of("endpoint").unwrap()).expect("endpoint is not a valid URL"),
+        http: reqwest::ClientBuilder::new()
+            .timeout(Duration::from_secs(5))
```

Comment:
is 5s enough for deleting a pool? This is also something that we might have to tweak on the core agent.

Reply:
We are not deleting the pool, but the spec. So I would assume, we don't have to wait? If we do, then for sure, yes, we need to increase it!

Reply:
Yes we do have to wait as things stand, unless we make the API async or something along those lines.